aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -166,259 +224,333 @@ module aptos_framework::fungible_asset {
+        metadata: Object<T>,
+        amount: u64,
+    ): FungibleAsset acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let owner_addr = signer::address_of(owner);
+        assert!(wallet_exists(owner_addr, metadata), error::invalid_argument(EINSUFFICIENT_BALANCE));
+        assert!(
+            ungated_transfer_allowed(owner_addr, metadata),
+            error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED),
+        );
+        let wallet = create_wallet_address_from_account(owner_addr, metadata);
+        withdraw_internal(wallet, amount)
+    }
+
+    /// Deposit fungible assets into an account.
+    public fun deposit(owner: address, fa: FungibleAsset) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        assert!(
+            !wallet_exists(owner, fa.metadata) ||
+                ungated_transfer_allowed(owner, fa.metadata),
+            error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED),
+        );
+        deposit_internal(owner, fa);
     }
 
-    /// Mint the `amount` of fungible asset.
+    /// Mint the specified `amount` of fungible asset.
     public fun mint(
         ref: &MintRef,
         amount: u64,
-    ): FungibleAsset acquires FungibleAssetMetadata {
+    ): FungibleAsset acquires FungibleAssetMetadata, FungibleAssetMetadataEvents {
         assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
-        let metadata = verify(&ref.metadata);
-        increase_supply(&ref.metadata, amount);
+        let metadata = ref.metadata;
+        increase_supply(&metadata, amount);
+
+        let events = borrow_global_mut<FungibleAssetMetadataEvents>(object::object_address(&metadata));
+        event::emit_event(&mut events.mint_events, MintEvent { amount });
+
         FungibleAsset {
             metadata,
             amount
         }
     }
 
+    /// Mint the specified `amount` of fungible asset to a destination address.
+    public fun mint_to(
+        ref: &MintRef,
+        to: address,
+        amount: u64,
+    ) acquires FungibleAssetMetadata, FungibleAssetMetadataEvents, FungibleAssetWallet, FungibleAssetWalletEvents {
+        deposit(to, mint(ref, amount));
+    }
+
     /// Enable/disable the direct transfer of fungible asset.
     public fun set_ungated_transfer(
         ref: &TransferRef,
-        wallet: &Object<FungibleAssetWallet>,
+        account: address,
         allow: bool,
-    ) acquires FungibleAssetWallet {
-        assert!(
-            &ref.metadata == &metadata_from_wallet(wallet),
-            error::invalid_argument(ETRANSFER_REF_AND_WALLET_MISMATCH)
-        );
-        borrow_fungible_asset_mut(wallet).allow_ungated_transfer = allow;
+    ) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let metadata = ref.metadata;
+        // Create the account in case the transfer ref owner wants to disable ungated transfer for accounts that do not
+        // have a wallet yet.
+        if (!wallet_exists(account, metadata)) {
```

Comment:
make this a function called `ensure_wallet_exist`

Reply:
Updated